### PR TITLE
Aligning syntax with MockK's

### PR DESCRIPTION
**Pull request by alexandrefavre4**
## Summary

This pull request aims to align Mockative syntax more closely with MockK's. The changes introduced here will help reduce friction for developers transitioning from MockK to Mockative.

## Proposed changes

**1. Infix notation support (backwards compatible)**

Enabled infix notation for `invokes`, `returns` and `throws`.

```kotlin
coEvery { api.getUsers() } returns users
```

**2. Expectation parameters as `verify` arguments (breaking change)**

Moved expectation parameters within `verify` arguments instead of chaining, aligning verification with MockK's approach. Also, developers can easily forget to chain `wasInvoked()` which will cause the `verify` block to be ignored.

```kotlin
verify { /* ... */ }
verify(exactly = never) { /* ... */ }
coVerify(atLeast = once) { /* ... */ }
```

**3. Replacement of `doesNothing` with `justRun` (breaking change)**

Replaced `doesNothing` with `justRun` for conciseness and to align with MockK's terminology.

```kotlin
justRun { user.deactivated = true }
```

**4. Renaming of `verifyNoUnverifiedExpectations` and `verifyNoUnmetExpectations` (breaking change)**

Renamed `verifyNoUnverifiedExpectations` to `confirmVerified` and `verifyNoUnmetExpectations` to `checkUnnecessaryStub` to align with MockK's terminology.

```kotlin
@AfterEach
fun afterEach() {
  confirmVerified(api)
  checkUnnecessaryStub(api)
}
```

### CODE REVIEW

**Comment by Nillerr** [comment]
Thank you for your suggestion @alexandrefavre4!

Truth be told I need to think about whether I like aligning even more with MockK. There are parts of MockK's API I don't think are as intuitive as the APIs I provide in Mockative, examples of which you have provided in the description of this PR. I don't necessarily think it has been an overall positive thing to align with MockK when I introduced the `every` function in Mockative 2, as the previous implementation of Mockative did have a substantial advantage in the sense that it doesn't need to do any fancy tricks to support its API. 

It would be possible to release a `mockative-mockk` module that provides MockK aliases for many of the APIs in Mockative.

In summary, now that I'm resuming maintenance of this project, I need to determine if I want Mockative to stand on its own as a competitor to MockK on the JVM, or whether I want it to be more like "MockK for KMP".

**Comment by vitorhugods** [comment]
Just my two cents about `verify(exactly = x)`.

When migrating from Mockative 1.x to Mockative 2.x, I used tons of Find/Replace + RegEx to do it. 
I slowly realised that in many places there was missing a `.wasInvoked`, so there was no assertion at all.

So I'm here just to testify that: yes, unfortunately it can be forgotten.

----

Other than that, I really don't care much about the differences between MockK and Mockative. Having independent approaches might be best than just blindly following someone else's footsteps.

I'd just like to avoid pitfalls like this whenever possible.

**Comment by Nillerr** [comment]
Thank you for your input @vitorhugods

It's a good point that one could forget to call `verify`, and a great argument for considering changing it. There is a way where I could retain the current API, but also make `verify { ... }` without `.wasInvoked()` implicitly be `wasInvoked(atLeast = once)`. I'll have a look at addressing that case in an upcoming release!

---

Now, while we're on the subject of the `verify { ... }` function I figured I wanted to share my thoughts through my experience over the past few years...

Regardless of what the documentation currently states I believe that you don't need to call `verify { ... }` as long as:

1. You're not using `stubsUnitByDefault = true`
2. You call `verifyNoUnmetExpectations` after every test
3. I introduce a way to mock an exact number of invocations (e.g. `returnsOnly(...)`)

A test of a function in the SUT returning `Unit` would thus look like this:

```kotlin
@Test
fun myTest() {
    // Given
    every { api.star("mockative") }
        .returnsOnly(Unit)

    // When
    val result = runCatching { service.star("mockative") }

    // Then
    assertSuccess(result)
}

@AfterEach
fun afterEach() {
    veryNoUnmetExpectations(this)
}
```

And any calls that haven't been stubbed, as well as any stubbed calls that haven't been called, would now produce an error.

Now, I realize that this differs from patterns that are often taught in books, articles, and tutorials, where using a `verify` function is often encouraged. Through my experience, I have simply come to learn that I never need the call to the `verify` function as long as I use `verifyNoUnmetExpectations` (or alternatives from MockK or Mockito).

Because of this, I am torn between whether Mockative should be its own thing, trying to rethink what a JVM mocking framework should be (e.g. deprecating and later removing the `verify { ... }` function to promote a different way of thinking), or whether it should "just" be "MocKK for KMP". Since Mockative is already released, and has a significant number of stars (and presumably usages), I can't help but feel an obligation to support an API that feels familiar to consumers of the framework, regardless of whether it aligns with my personal preferences and beliefs of "correctness" 🤷🏻‍♂️ 

From a technology point of view, Mockative is already fundamentally different from MockK and Mockito since Mockative uses code generation to provide mocks, whereas MockK and Mockito use runtime bytecode generation.

**Comment by alexandrefavre4** [comment]
Totally understand your point @Nillerr. Looking at the API, I got the feeling that Mockative wanted to position itself as MockK for KMP, that is why I thought that closing the gap could help consumers as it would for the company I work for.

@vitorhugods I definitely agree with you, Mockative shouldn't blindly follow MockK footsteps. Coming back to the proposed changes, I think that points 1 and 2 still add value other than aligning with MockK. For point 1, some people find infix notation more readable. It's definitely a low hanging fruit to me, as it doesn't impact the current API. For point 2, I think we all agree that `verify(...)` has some benefits over `wasInvoked(...)` (e.g. can't forget `wasInvoked`, conciseness). For points 3 and 4, main benefit is to have the API more closely aligned with MockK. Now, I personally find `checkUnnecessaryStub`/`confirmVerified` clearer than `verifyNoUnmetExpectations`/`verifyNoUnverifiedExpectations` (too many negations 😵‍💫) and have a slight preference for `justRun` over `doesNothing`, but I guess that's a matter of personal preferences.

I propose you that we close this pull request and that we proceed with the following steps:
- We skip points 3 and 4 for now - Up to you to consider these changes according to your vision for Mockative
- I create another pull request for adding infix notation support
- I create another pull request and keep only changes for point 2 and try to see how we could retain the existing API

---

About your thoughts on `verify`, we could indeed skip `verify` statements when these conditions are met. However I think that `verify` helps the test to be more explicit (e.g. I want to this specific function call to be made) and errors are also more explicit than getting MissingStubSomething exceptions. Finally it also helps in readability by having the verification done in the `then` block, next to the assertions.

**Comment by alexandrefavre4** [comment]
Infix notation support has been added (see https://github.com/mockative/mockative/pull/100). I'll look into point 2 when I have some time.

**Comment by alexandrefavre4** [comment]
Infix notation support has been added (see https://github.com/mockative/mockative/pull/100). I'll look into point 2 when I have some time.
